subsys/bluetooth/services/bas/bas.c

```diff
@@ -131,4 +169,66 @@ const struct bt_gatt_attr *bt_bas_get_bas_attr(uint16_t index)
 	return NULL;
 }
 
+#if defined(CONFIG_BT_BAS_BCS)
+
+static void bcs_indicate_cb(struct bt_conn *conn, struct bt_gatt_indicate_params *params,
+			    uint8_t err)
+{
+	if (err != 0) {
+		LOG_DBG("BCS Indication failed with error %d\n", err);
+	} else {
+		LOG_DBG("BCS Indication sent successfully\n");
+	}
+}
+
+static void bt_bas_bcs_update_battery_critical_status(void)
+{
+	uint8_t err;
+
+	/* Indicate all connections */
+	bcs_ind_params.attr = &bas.attrs[BT_BAS_BATTERY_CRITICAL_STATUS_CHAR_HANDLE];
+	bcs_ind_params.data = &battery_critical_status;
+	bcs_ind_params.len = sizeof(battery_critical_status);
+	bcs_ind_params.func = bcs_indicate_cb;
+	err = bt_gatt_indicate(NULL, &bcs_ind_params);
+	if (err) {
+		LOG_DBG("Failed to send ind to all connections (err %d)\n", err);
+	}
+}
+
+void bt_bas_bcs_set_battery_critical_state(bool critical_state)
+{
+	bool current_state = (battery_critical_status & BCS_BATTERY_CRITICAL_STATE) != 0;
+
+	if (current_state == critical_state) {
+		LOG_INF("Already battery_critical_state is %d\n", critical_state);
+		return;
+	}
+
+	if (critical_state) {
+		battery_critical_status |= BCS_BATTERY_CRITICAL_STATE;
+	} else {
+		battery_critical_status &= ~BCS_BATTERY_CRITICAL_STATE;
+	}
+	bt_bas_bcs_update_battery_critical_status();
+}
+
+void bt_bas_bcs_set_immediate_service_required(bool service_required)
+{
+	bool current_state = (battery_critical_status & BCS_IMMEDIATE_SERVICE_REQUIRED) != 0;
+
+	if (current_state == service_required) {
+		LOG_INF("Already immediate_service_required is %d\n", service_required);
+		return;
+	}
+
+	if (service_required) {
+		battery_critical_status |= BCS_IMMEDIATE_SERVICE_REQUIRED;
+	} else {
+		battery_critical_status &= ~BCS_IMMEDIATE_SERVICE_REQUIRED;
+	}
+	bt_bas_bcs_update_battery_critical_status();
+}
+#endif /* CONFIG_BT_BAS_BCS */
```

Comment:
I don't understand your approach. 

Somethings are defined in `bas.c` and only used in `bas_bls.c` and vice versa. 

The Battery critical status is not part of BLS (even though it depends on it), and should be `bas.c` in perhaps `bas_bcs.c`

Reply:
As per spec, Battery Critical power state bit should be set based on Battery Level Status Power State field and Service Required bits is set to True based on the Service Required bit in the Additional Status field of the Battery Level Status 

3.4.1.1 Battery Critical Status field

The value of the Critical Power State bit shall be set to True if the value of the Battery Charge Level bits is
set to Critical in the Power State field of the Battery Level Status characteristic if implemented. Otherwise,
the Critical Power State bit shall be set to False.

The value of the Immediate Service Required bit shall be set to True if the value of the Service Required
bits is set to True in the Additional Status field of the Battery Level Status characteristic if implemented.
Otherwise, the Immediate Service Required bit shall be set to False

Reply:
Right, but we really shouldn't have half of the implementation of the characteristic in one file and the rest in another file (especially since those clearly aren't even integrated: https://github.com/zephyrproject-rtos/zephyr/pull/79966#discussion_r1804725431)